[macOS] Add a version of `-_setCustomSwipeViewsTopContentInset:` that supports all 4 rect edges

whsieh opened the pull request:
#### c2abdb76db274ba599d89d60075d88eb268e80c0
<pre>
[macOS] Add a version of `-_setCustomSwipeViewsTopContentInset:` that supports all 4 rect edges
<a href="https://bugs.webkit.org/show_bug.cgi?id=290731">https://bugs.webkit.org/show_bug.cgi?id=290731</a>
<a href="https://rdar.apple.com/144614910">rdar://144614910</a>

Reviewed by Abrar Rahman Protyasha.

Add an alternate version of `-_setCustomSwipeViewsTopContentInset:` that takes an `NSEdgeInsets`
struct instead of just a single value denoting the top content inset, and reimplement the existing
method in terms of the new method.

See below for more details.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(coreBoxExtentsFromEdgeInsets):
(-[WKWebView _setObscuredContentInsets:immediate:]):

Pull logic to convert `NSEdgeInsets` into a `WebCore::FloatBoxExtent` out into a separate static
helper, so that I can reuse it below.

(-[WKWebView _setCustomSwipeViewsTopContentInset:]):

Reimplement this in terms of `-_setCustomSwipeViewsObscuredContentInsets:`.

(-[WKWebView _setCustomSwipeViewsObscuredContentInsets:]):
* Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewTestingMac.mm:
(-[WKWebView _windowRelativeBoundsForCustomSwipeViewsForTesting]):

Add a simple testing hook to ask for the window-relative bounds for custom swipe views. This isn&apos;t
really exposed in any other sensible way for API tests, so I fell back to adding a new testing hook
instead.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::windowRelativeBoundsForCustomSwipeViews const):

Add support for all 4 rect edges here. Instead of just shrinking `swipeArea` by the top inset, we
now shrink `swipeArea` by all 4 sides (note that this is in AppKit window coordinates, so the origin
is in the bottom left corner).

(WebKit::ViewGestureController::beginSwipeGesture):

This previously re-expanded the `windowRelativeBoundsForCustomSwipeViews` by the top content inset;
we adjust this to do the same for all 4 rect edges, by unapplying the shrinking by obscured insets
above.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::windowRelativeBoundsForCustomSwipeViews const):
(WebKit::WebViewImpl::customSwipeViewsObscuredContentInsets const):
(WebKit::WebViewImpl::protectedGestureController const):

Add a `protectedGestureController` const getter that returns a `RefPtr`.

(WebKit::WebViewImpl::setCustomSwipeViewsObscuredContentInsets):
(WebKit::WebViewImpl::setCustomSwipeViewsTopContentInset): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/CustomSwipeViewTests.mm: Copied from Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h.
(TestWebKitAPI::TEST(CustomSwipeViewTests, WindowRelativeBoundsForCustomSwipeViews)):

Add a new simple API test to exercise this SPI, using the new test helper property.

Canonical link: <a href="https://commits.webkit.org/292988@main">https://commits.webkit.org/292988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ddb7af194e81d2da5259af72e51e884947925ff0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7459 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102724 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48147 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17537 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74366 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31552 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13282 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54715 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13054 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6162 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47589 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6238 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104725 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24698 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18011 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83414 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25070 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84404 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82843 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20863 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27396 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18293 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24659 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24481 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27795 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26055 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
```diff
@@ -311,7 +311,11 @@ static float maximumRectangleComponentDelta(FloatRect a, FloatRect b)
     for (const auto& view : m_customSwipeViews)
         swipeArea.unite([view convertRect:[view bounds] toView:nil]);
-    swipeArea.setHeight(swipeArea.height() - m_customSwipeViewsTopContentInset);
+    swipeArea.move(m_customSwipeViewsObscuredContentInsets.left(), m_customSwipeViewsObscuredContentInsets.bottom());
```
would we not want to move to the top left edge?

aprotyas replied:
oh, my bad, bottom of the insets. That seems correct

whsieh replied:
Ah, yeah — this is because bottom left corner is the origin of `NSWindow` (as opposed to top left in UIKit, and on the web)

whsieh committed:
[macOS] Add a version of `-_setCustomSwipeViewsTopContentInset:` that supports all 4 rect edges
https://bugs.webkit.org/show_bug.cgi?id=290731
rdar://144614910

Reviewed by Abrar Rahman Protyasha.

Add an alternate version of `-_setCustomSwipeViewsTopContentInset:` that takes an `NSEdgeInsets`
struct instead of just a single value denoting the top content inset, and reimplement the existing
method in terms of the new method.

See below for more details.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(coreBoxExtentsFromEdgeInsets):
(-[WKWebView _setObscuredContentInsets:immediate:]):

Pull logic to convert `NSEdgeInsets` into a `WebCore::FloatBoxExtent` out into a separate static
helper, so that I can reuse it below.

(-[WKWebView _setCustomSwipeViewsTopContentInset:]):

Reimplement this in terms of `-_setCustomSwipeViewsObscuredContentInsets:`.

(-[WKWebView _setCustomSwipeViewsObscuredContentInsets:]):
* Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewTestingMac.mm:
(-[WKWebView _windowRelativeBoundsForCustomSwipeViewsForTesting]):

Add a simple testing hook to ask for the window-relative bounds for custom swipe views. This isn't
really exposed in any other sensible way for API tests, so I fell back to adding a new testing hook
instead.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::windowRelativeBoundsForCustomSwipeViews const):

Add support for all 4 rect edges here. Instead of just shrinking `swipeArea` by the top inset, we
now shrink `swipeArea` by all 4 sides (note that this is in AppKit window coordinates, so the origin
is in the bottom left corner).

(WebKit::ViewGestureController::beginSwipeGesture):

This previously re-expanded the `windowRelativeBoundsForCustomSwipeViews` by the top content inset;
we adjust this to do the same for all 4 rect edges, by unapplying the shrinking by obscured insets
above.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::windowRelativeBoundsForCustomSwipeViews const):
(WebKit::WebViewImpl::customSwipeViewsObscuredContentInsets const):
(WebKit::WebViewImpl::protectedGestureController const):

Add a `protectedGestureController` const getter that returns a `RefPtr`.

(WebKit::WebViewImpl::setCustomSwipeViewsObscuredContentInsets):
(WebKit::WebViewImpl::setCustomSwipeViewsTopContentInset): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/CustomSwipeViewTests.mm: Copied from Source/WebKit/UIProcess/API/mac/WKWebViewPrivateForTestingMac.h.
(TestWebKitAPI::TEST(CustomSwipeViewTests, WindowRelativeBoundsForCustomSwipeViews)):

Add a new simple API test to exercise this SPI, using the new test helper property.

Canonical link: https://commits.webkit.org/292988@main
